p11-kit/proxy.c

```diff
@@ -279,29 +278,46 @@ proxy_list_slots (Proxy *py, Mapping *mappings, unsigned int n_mappings)
 
 		if (count > 0) {
 			Mapping *new_mappings;
+			bool *is_new_mapping;
 
+			is_new_mapping = calloc (count, sizeof(bool));
+			return_val_if_fail (is_new_mapping != NULL, CKR_HOST_MEMORY);
 			new_mappings = realloc (py->mappings, sizeof (Mapping) * (py->n_mappings + count));
 			return_val_if_fail (new_mappings != NULL, CKR_HOST_MEMORY);
 			py->mappings = new_mappings;
 
-			/* And now add a mapping for each of those slots */
+			/* Reuse the existing mapping if any */
 			for (i = 0; i < count; ++i) {
-				/* Reuse the existing mapping if any */
 				for (j = 0; j < n_mappings; ++j) {
 					/* cppcheck-suppress nullPointer symbolName=mappings */
 					/* false-positive: https://trac.cppcheck.net/ticket/9573 */
 					if (mappings[j].funcs == funcs &&
-					    mappings[j].real_slot == slots[i])
+					    mappings[j].real_slot == slots[i]) {
+						py->mappings[py->n_mappings].funcs = funcs;
+						py->mappings[py->n_mappings].real_slot = slots[i];
+						py->mappings[py->n_mappings].wrap_slot =
+							mappings[j].wrap_slot;
+						++py->n_mappings;
 						break;
+					}
+				}
+				if (n_mappings == 0 || j == n_mappings)
+					is_new_mapping[i] = true;
+			}
+
+			/* And now add a mapping for each new slot */
+			for (i = 0; i < count; ++i) {
+				if (is_new_mapping[i]) {
```

Comment:
nit: I suspect this loop can be simpler if you record slot IDs instead of boolean flags.

Reply:
Makes sense, I'll change it.

Reply:
Changed it to store the ID's.